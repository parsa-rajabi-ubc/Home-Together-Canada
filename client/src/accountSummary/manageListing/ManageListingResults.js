/**
 *
 * @Author:     Parsa Rajabi
 * @Created:    2021.6.9
 *
 * @Description: Manage Listing Cards
 *
 */

import React, {useState} from "react";
import PropTypes from "prop-types";
import {MdDeleteForever} from "react-icons/md";
import Modal from 'react-modal';
import {TAB_LABELS} from "./ManageListingTabs";
import MemberListingCard from "../../searchServicesClassifieds/listingCards/MemberListingCard";
import Paginate from "../../common/forms/Paginate";
import {customModalStyles} from "../../css/ModelCSSUtil";


const MANAGE_LISTING_TEXT = {
    TITLE: "Manage Listings",
    DESCRIPTION: "Use the tabs below to view your listings based on their status.",
    NO_LISTING: "No ",
    CONFIRM_DELETE: "Are you sure you want to delete listing \"",
    DELETE: "Delete",
    CANCEL: "Cancel"
}

const NUM_RESULTS = 7;
Modal.setAppElement('body')

function ManageListingResults(props) {
    const {
        onSubmit,
        viewableListingData,
        activeTab,
        setViewableListingData
    } = props;

    const [isModalOpen, setIsModalOpen] = useState(false);
    const [listingID, setListingID] = useState();
    const [listingTitle, setListingTitle] = useState();

    function openModal() {
        setIsModalOpen(true);
    }

    function closeModal() {
        setIsModalOpen(false);
    }

    function deleteListing() {
        closeModal();
        onSubmit(listingID);
    }

    const onClickDecision = (listingID, listingTitle) => {
        setListingID(listingID);
        setListingTitle(listingTitle);
        setViewableListingData(listingTitle);
        openModal();
    }

    //TODO: Refactor to handle business cards as well
    function displayListingCards() {
        const listingCards = viewableListingData.map(
            (listing) =>
                <div className={"flex"} key={listing.id}>
                    <section className={"w-full"}>
                        <MemberListingCard
                            title={listing.title}
                            monthlyCost={listing.monthlyCost}
                            petFriendly={listing.petFriendly}
                            smokeFriendly={listing.smokeFriendly}
                            shortDescription={listing.shortDescription}
                            datePosted={listing.createdAt}/>
                    </section>

                    {/* only show delete button (garbage can) when the listing status is LIVE */}
                    {activeTab === TAB_LABELS.LIVE &&
                    <section className={"my-8"}>
                        <MdDeleteForever
                            color="#DB4437"
                            size="40"
                            onClick={() => onClickDecision(listing.id, listing.title)}
                        />
                    </section>
                    }
                </div>
        );

        return <Paginate data={listingCards} resultsPerPage={NUM_RESULTS}/>
    }

    return (
        <div>
            {viewableListingData && displayListingCards()}

            {viewableListingData && <Modal
                isOpen={isModalOpen}
                onRequestClose={closeModal}
                style={customModalStyles}
                contentLabel="Confirm Decision for Listing"
            >

                <h1 className={"text-gray-800 h1"}>{MANAGE_LISTING_TEXT.CONFIRM_DELETE}{listingTitle + "\"?"} </h1>
                <section className={"flex pt-10 space-x-10"}>
                    <button type={"button"} className="py-2 w-1/2 bg-gray-500 btn" onClick={closeModal}>
                        {MANAGE_LISTING_TEXT.CANCEL}
                    </button>
                    <button
                        type={"submit"}
                        className={"btn-red btn w-1/2 py-2"}
                        onClick={deleteListing}
                        value={MANAGE_LISTING_TEXT.DELETE}
                    >
                        {MANAGE_LISTING_TEXT.DELETE}
                    </button>
                </section>
            </Modal>
            }
        </div>
    );
}

ManageListingResults.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    viewableListingData: PropTypes.array,
    activeTab: PropTypes.string,
    setViewableListingData: PropTypes.func,
};

export default ManageListingResults;