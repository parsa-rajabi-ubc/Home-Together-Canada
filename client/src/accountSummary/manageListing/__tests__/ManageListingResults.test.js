/**
 * @Author:     Parsa Rajabi
 * @Created:    2021.06.13
 *
 * @Description: Manage Listing Test
 *
 */
import React from 'react';
import renderer from 'react-test-renderer'
import ManageListingResults from "../ManageListingResults";
import {BrowserRouter} from "react-router-dom";


jest.mock("uuid", () => ({
            v4: () => "00000000-0000-0000-0000-000000000000"
        }
    )
);

jest.mock('react-redux', () => ({
    connect: () => {
        return (component) => {
            return component
        };
    }
}));

describe('ManageListing', () => {
    describe('Snapshot test', () => {
        it("should render correctly regardless of properties", () => {
            // given
            const props = {
                onSubmit: jest.fn(),
                listingData: [{
                    "id": 51,
                    "uid": 5,
                    "title": "Empty Room in Condo 7",
                    "shortDescription": "Looking for an easy going roommate",
                    "fullDescription": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
                    "monthlyCost": 600,
                    "numBed": 2,
                    "numBath": 1,
                    "petFriendly": true,
                    "smokeFriendly": false,
                    "utilitiesIncluded": true,
                    "addressLine1": "2271 Harvey Ave",
                    "city": "Kelowna",
                    "province": "BC",
                    "postalCode": "V1Y6H2",
                    "isClassified": false,
                    "createdAt": "2021-06-14T00:42:47.000Z",
                    "updatedAt": "2021-06-14T00:42:47.000Z",
                    "categoryName": "Members with Homes to Share",
                    "images": []
                }],
                activeTab: "Live",
                accountType: 'member'
            }

            // when
            const component = renderer.create(<BrowserRouter><ManageListingResults {...props} /></BrowserRouter>).toJSON();

            // then
            expect(component).toMatchSnapshot();
        });
    })
})