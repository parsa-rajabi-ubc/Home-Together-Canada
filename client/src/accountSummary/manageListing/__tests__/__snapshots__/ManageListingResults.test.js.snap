// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageListing Snapshot test should render correctly regardless of properties 1`] = `<div />`;
